libmcount/dynamic.c

```diff
@@ -327,12 +354,15 @@ static void prepare_dynamic_update(struct uftrace_sym_info *sinfo, bool needs_mo
 	};
 	int hash_size = sinfo->exec_map->mod->symtab.nr_sym * 3 / 4;
 
-	if (needs_modules)
-		hash_size *= 2;
-
-	code_hmap = hashmap_create(hash_size, hashmap_ptr_hash, hashmap_ptr_equals);
+	if (code_hmap && !needs_modules) /* already prepared */
+		return;
+	if (!code_hmap) { /* nothing loaded */
+		if (needs_modules)
+			hash_size *= 2;
+		code_hmap = hashmap_create(hash_size, hashmap_ptr_hash, hashmap_ptr_equals);
+	}
```

Comment:
What happens if the first update is for main only and the next one needs modules.  I think hash size would be kept only for main, right?

Reply:
Yes I spotted that and looked strange to me as well. I'll fix it.